### ISSUE TITLE
Title: optimizer=auto incorrectly sets warmup_bias_lr=0.0 when MuSGD is selected

### ISSUE THREAD

**Issue by songjiahao-wq** [ultralytics/ultralytics#23637]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

When optimizer=auto resolves to MuSGD (iterations > 10000), warmup_bias_lr is incorrectly forced to 0.0. This setting is intended for Adam-family optimizers only, but
  currently applies unconditionally to both AdamW and MuSGD.

  Location

  ultralytics/engine/trainer.py, build_optimizer() method:

  # Line 951-952
  name, lr, momentum = ("MuSGD", 0.01, 0.9) if iterations > 10000 else ("AdamW", lr_fit, 0.9)
  self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam  <-- always executes

  The comment says "no higher than 0.01 for Adam", but this line executes regardless of whether MuSGD or AdamW is selected.

  Impact

  - SGD/MuSGD benefits from a non-zero warmup_bias_lr (default 0.1) during warmup to stabilize early training.
  - With warmup_bias_lr=0.0, bias parameters receive zero learning rate at the start of warmup, which can degrade final accuracy.
  - This is not visible in logs because print_args() is called at line 156 (before build_optimizer), so the printed warmup_bias_lr=0.1 does not reflect the actual value used
  during training.
  - Manually specifying optimizer=MuSGD avoids this issue and produces better results with the same lr and momentum.

  To Reproduce

  from ultralytics import YOLO

  model = YOLO("yolo11s-seg.pt")

  # Run 1: auto → MuSGD, warmup_bias_lr silently set to 0.0
  model.train(data="my_data.yaml", epochs=400, optimizer="auto")

  # Run 2: manual MuSGD, warmup_bias_lr stays at 0.1
  model.train(data="my_data.yaml", epochs=400, optimizer="MuSGD")

  # Run 2 produces noticeably higher accuracy than Run 1,
  # despite both using MuSGD with the same lr and momentum.

  Suggested Fix

  Move warmup_bias_lr = 0.0 inside the AdamW branch only:

  if iterations > 10000:
      name, lr, momentum = "MuSGD", 0.01, 0.9
  else:
      name, lr, momentum = "AdamW", lr_fit, 0.9
      self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam

  Environment

  - Ultralytics: 8.4.13
  - Python: 3.11
  - PyTorch: 2.9.1+cu128
  - GPU: NVIDIA A100-PCIE-40GB

  ---

### Environment

(YOLO) root@sx-dms0925:/data/lost+found/sjh/line_seg# yolo check
Ultralytics 8.4.13 🚀 Python-3.11.14 torch-2.9.1+cu128 CUDA:0 (NVIDIA A100-PCIE-40GB, 40444MiB)
Setup complete ✅ (96 CPUs, 754.4 GB RAM, 409.7/438.9 GB disk)

OS                     Linux-5.15.0-156-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.11.14
Install                pip
Path                   /root/miniforge3/envs/YOLO/lib/python3.11/site-packages/ultralytics
RAM                    754.35 GB
Disk                   409.7/438.9 GB
CPU                    Intel Xeon Gold 6248R CPU @ 3.00GHz
CPU count              96
GPU                    NVIDIA A100-PCIE-40GB, 40444MiB
GPU count              8
CUDA                   12.8

numpy                  ✅ 2.2.0>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 12.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.1>=1.8.0
torch                  ✅ 2.9.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1>=0.9.0
psutil                 ✅ 7.1.3>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

```
from ultralytics import settings
settings["tensorboard"]=True
from ultralytics import YOLO

from pathlib import Path
import os
import torch
import gc
import sys
import argparse

# from ultralytics import RTDETR


def parse_args():
    parser = argparse.ArgumentParser(description="YOLO Training Script")
    parser.add_argument(
        "--exp_name",
        type=str,
        default=None,
        help="训练实验名称（保存目录名），如 yolo11x_text_v1"
    )
    return parser.parse_args()

if __name__ == '__main__':
    
    args = parse_args()

    # 加载模型
    weights = "./DATA/yolo26s-seg.pt"
    yaml = "ultralytics/cfg/models/26/yolo26-seg.yaml"
    model = YOLO(weights)
    # model = YOLO(yaml).load(weights)  # 从头开始构建新模型
    # model = YOLO("runs/detect/v5n9/weights/best.pt")  # 加载预训练模型（推荐用于训练）
    results = model.train(
        data="DATA/Complete_line/Complete_line_v1.0.yaml",
        epochs=400,
        batch=16,
        imgsz=640,
        workers=14,
        # name=args.exp_name or Path(model.cfg).stem,
        name=args.exp_name or Path(weights).stem, 
        multi_scale=0.25,
        save_period=100,
        # inner_iou=True,
        # inner_iou_version=2,
        # use_elongated_loss=True,
        # project="./runs",
        # device=[6]
        # lr0=0.005,
        optimizer= 'auto' # (str) optimizer: SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, or auto
    )
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @songjiahao-wq, thank you for the detailed report and for your interest in Ultralytics 🚀! This is an automated response to help move things along quickly—an Ultralytics engineer will also assist soon 🙏

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. You’ve already included an MRE (thank you!)—if you can, please also add the smallest possible snippet that reproduces the `optimizer="auto"` → `MuSGD` behavior and clearly shows the resulting `warmup_bias_lr` value used during training (not only what prints before optimizer build) 🔍

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by songjiahao-wq** [comment]
```
        if name == "auto":
            LOGGER.info(
                f"{colorstr('optimizer:')} 'optimizer=auto' found, "
                f"ignoring 'lr0={self.args.lr0}' and 'momentum={self.args.momentum}' and "
                f"determining best 'optimizer', 'lr0' and 'momentum' automatically... "
            )
            nc = self.data.get("nc", 10)  # number of classes
            lr_fit = round(0.002 * 5 / (4 + nc), 6)  # lr0 fit equation to 6 decimal places
            name, lr, momentum = ("MuSGD", 0.01, 0.9) if iterations > 10000 else ("AdamW", lr_fit, 0.9)
            self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam
```
 当我设置 optimizer=auto 时，ultralytics 会自动选择优化器：
  - 数据量大（iterations > 10000）→ 选 MuSGD
  - 数据量小 → 选 AdamW

  代码里有一行 self.args.warmup_bias_lr = 0.0，本意是给 AdamW 用的（Adam 类优化器自带自适应学习率，warmup 阶段 bias 学习率不需要额外设置，设高了反而会震荡）。

  但这行代码写在了选择优化器之后、不分 MuSGD 还是 AdamW 都会执行。 导致即使 auto 选了 MuSGD，warmup_bias_lr 也被强制改成了 0.0。

  MuSGD/SGD 是需要 warmup_bias_lr=0.1 的，被改成 0.0 后 warmup 阶段 bias 参数完全不更新，影响训练效果，最终精度下降。

  而且这个修改发生在日志打印之后，日志里显示的还是 0.1，用户完全看不出来实际用的是 0.0。

**Comment by Y-T-G** [comment]
@Laughing-q Is this a bug?

**Comment by Y-T-G** [comment]
@songjiahao-wq Can you open a PR?

**Comment by songjiahao-wq** [comment]
For the training of my dataset, if warmup_bias_lr is set to 0, the following situation with extremely poor accuracy will occur.

<img width="1027" height="189" alt="Image" src="https://github.com/user-attachments/assets/1bf2a745-1819-4244-a686-3bf3c459050b" />

warmup_bias_lr is set to 0.1

<img width="1005" height="187" alt="Image" src="https://github.com/user-attachments/assets/160747bf-ce16-4867-aac8-ddbd8abd0252" />

**Comment by Laughing-q** [comment]
@songjiahao-wq @Y-T-G I think we kept this for a long time for SGD as well in previous versions before we introduced MuSGD so I kept the same logic there.
Also `warmup_bias_lr==0` means in warm up period the lr would be increasing from 0 to the set `lr0`. It's not `warmup_bias_lr=0` the lr would stay 0 during warmup period. And `warmup_bias_lr=0.1`, means the lr would be decreasing from 0.1 to `lr0`.

**Comment by Laughing-q** [comment]
@songjiahao-wq in your case probably at the beginning epochs with more bias changes would be more fit to your datasets so `warmup_bias_lr=0.1` with overall higher lr than starting from `warmup_bias_lr=0.0` would be better

**Comment by songjiahao-wq** [comment]
@Laughing-q Thanks for the explanation @Laughing-q.                                                                                                                                                    
                                                                                                                                                                                             
  I understand that warmup_bias_lr=0 means the bias lr increases from 0 to lr0 during warmup, not that it stays at 0. However, I'd like to raise two points:                                 
                                                                                                                                                                                             
  1. It does affect accuracy in practice.

  I ran two experiments with identical settings on the same dataset (1795 images, 400 epochs, seed=0), the only difference being optimizer=auto vs optimizer=MuSGD:

  - optimizer=auto → MuSGD(lr=0.01, momentum=0.9), warmup_bias_lr silently changed to 0.0 → lower accuracy
  - optimizer=MuSGD → MuSGD(lr=0.01, momentum=0.9), warmup_bias_lr stays at 0.1 → noticeably better accuracy

  So the forced warmup_bias_lr=0.0 does have a measurable negative impact, at least for MuSGD on small datasets.

  2. This override is invisible to users.

  print_args() is called at line 156, but self.args.warmup_bias_lr = 0.0 happens at line 952 inside build_optimizer(). So the log always prints warmup_bias_lr=0.1 even when the actual value
   used is 0.0. This makes it very hard to debug — I spent a long time figuring out why optimizer=auto and optimizer=MuSGD produced different results when all printed parameters looked
  identical.

  Even if this behavior is intentional, could we at least log the override so users are aware? For example:

  if name == "auto":
      ...
      self.args.warmup_bias_lr = 0.0
      LOGGER.info(f"optimizer=auto: setting warmup_bias_lr=0.0")

  Or move print_args() after build_optimizer() so the printed values reflect what is actually used

**Comment by songjiahao-wq** [comment]
> [@songjiahao-wq](https://github.com/songjiahao-wq) Can you open a PR?

I'm a beginner. I'll go through the details of the bug.

**Comment by songjiahao-wq** [comment]
@Y-T-G gh pr checkout 23642
https://github.com/ultralytics/ultralytics/pull/23642

**Comment by Br1an67** [comment]
I'd like to work on this. The fix moves the `warmup_bias_lr = 0.0` assignment into the AdamW-specific branch, preventing it from being incorrectly applied when MuSGD is auto-selected. PR: #23831

**Comment by glenn-jocher** [comment]
Thanks for opening [PR #23831](https://github.com/ultralytics/ultralytics/pull/23831); let’s keep review on the PR thread so everything stays in one place, and if possible please add a small regression test for the `optimizer=auto` → `MuSGD` path. ([github.com](https://github.com/ultralytics/ultralytics/pull/23831))
